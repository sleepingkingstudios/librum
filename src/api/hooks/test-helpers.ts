import type {
  Alert,
  AlertsContext,
} from '@alerts';
import type { Dispatch } from '@store';
import type { EffectOptions } from '../types';
import type {
  Response,
  UseQueryResult,
} from './types';

const alerts: AlertsContext = {
  alerts: [] as Alert[],
  dismissAlert: jest.fn(),
  dismissAllAlerts: jest.fn(),
  displayAlert: jest.fn(),
}
const dispatch: Dispatch = jest.fn();

export const defaultOptions: EffectOptions = {
  alerts,
  dispatch,
};

export const defaultResponse: Response = {
  hasData: false,
  hasError: false,
  isUninitialized: false,
  isLoading: false,
  isErrored: false,
  isFailure: false,
  isSuccess: false,
  status: 'unknown',
};

export const defaultResult: UseQueryResult = {
  isUninitialized: false,
  isLoading: false,
  isSuccess: false,
  isError: false,
  refetch: jest.fn(),
};
