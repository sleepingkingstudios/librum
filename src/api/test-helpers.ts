import { defaultTo } from 'lodash';

import type { DataObject } from '@utils/types';
import type {
  ApiError,
  FetchFailure,
  FetchSuccess,
} from './types';

interface IBuildFailureResponseOptions {
  data?: DataObject;
  error?: ApiError;
  status?: number;
}

interface IBuildSuccessResponseOptions {
  data?: DataObject;
}

export const buildFailureResponse = (
  options?: IBuildFailureResponseOptions
): FetchFailure => {
  if (options === null || options == undefined) {
    return { error: { data: {}, status: 400 }};
  }

  const data = defaultTo(options.data, null);
  const error = defaultTo(options.error, null);
  const status = defaultTo(options.status, 400);

  if (error === null) {
    return { error: { data: {}, status }};
  }

  if (data === null) {
    return {
      error: {
        status,
        data: {
          ok: false,
          error,
        },
      },
    };
  }

  return {
    error: {
      status,
      data: {
        ok: false,
        data,
        error,
      },
    },
  };
};

export const buildSuccessResponse = (
  options?: IBuildSuccessResponseOptions
): FetchSuccess => {
  if (options === null || options == undefined) {
    return { data: { ok: true } }
  }

  const data = defaultTo(options.data, null);

  if (data === null) {
    return { data: { ok: true } }
  }

  return { data: { data, ok: true } };
};

export const fetchErrorResponse: FetchFailure = {
  error: {
    error: 'something went wrong',
    status: 'FETCH_ERROR',
  },
};

export const serializedErrorResponse: FetchFailure = {
  error: { message: 'something went wrong' }
};
