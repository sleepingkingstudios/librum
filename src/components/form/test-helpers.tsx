import * as React from 'react';
import {
  Form,
  Formik,
  FormikValues,
} from 'formik';

import type { OnSubmit } from './types';

interface IFormWrapperProps {
  initialValues?: FormikValues;
  onSubmit?: OnSubmit;
  submitButton?: true | false;
}

const SubmitButton = (
  { submitButton }: { submitButton: true | false }
): JSX.Element => {
  if (!submitButton) { return null; }

  return (
    <button type="submit">Submit</button>
  );
};

export const formWrapper = (
  { initialValues, onSubmit, submitButton }: IFormWrapperProps) => {
  const FormWrapper = (
    { children }: { children: React.ReactNode }
  ): JSX.Element => (
    <Formik initialValues={initialValues || {}} onSubmit={onSubmit}>
      <Form>
        { children }

        <SubmitButton submitButton={submitButton} />
      </Form>
    </Formik>
  );

  return FormWrapper;
};
