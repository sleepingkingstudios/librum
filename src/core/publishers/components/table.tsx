import * as React from 'react';
import { Link } from 'react-router-dom';

import { DataTable } from '@components/data-table';
import type {
  DataTableColumn,
  DataTableData,
} from '@components/data-table';
import { ExternalLink } from '@components/external-link';

type MaybePublisher = {
  name?: string,
  slug?: string,
  website?: string,
};

const PublisherLink = ({ data }: { data: MaybePublisher }): JSX.Element => (
  <Link to={`/publishers/${data.slug}`}>{ data.name }</Link>
);

const WebsiteLink = ({ data }: { data: MaybePublisher }): JSX.Element => (
  <ExternalLink to={data.website} />
);

const columns: DataTableColumn[] = [
  {
    name: 'name',
    value: PublisherLink,
  },
  {
    name: 'website',
    value: WebsiteLink,
  },
  {
    label: false,
    name: 'actions',
  }
];

export const PublishersTable = ({
  data,
}: {
  data: DataTableData,
}): JSX.Element => (
  <DataTable
    collapse
    columns={columns}
    data={data}
    name="publishers"
    scope="core"
  />
);
