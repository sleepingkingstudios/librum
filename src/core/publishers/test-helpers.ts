import type { Publisher } from './types';

export const publishers: Publisher[] = [
  {
    id: 'e7c781df-ed94-45a5-90ee-94d28cc8f2dc',
    name: 'Crystal Sphere Publications',
    slug: 'crystal-sphere-publications',
    website: 'www.example.com/crystal-sphere-publications',
  },
  {
    id: 'eb79903b-3b73-420e-8e5a-604f27138260',
    name: 'Flumph Free Press',
    slug: 'flumph-free-press',
    website: 'www.example.com/flumph-free-press',
  },
  {
    id: 'dc40f9c7-69e5-47b5-bcde-d062bc6c44b1',
    name: 'Spelljammer Monthly',
    slug: 'spelljammer-monthly',
    website: 'www.example.com/spelljammer-monthly',
  },
];
