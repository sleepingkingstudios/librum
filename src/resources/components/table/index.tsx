import * as React from 'react';

import type { Response } from '@api';
import type { DataTableData } from '@components/data-table';
import { LoadingOverlay } from '@components/loading-overlay';
import type { ConfiguredDataTable } from '@resources/types';
import { titleCase } from '@utils/text';

type ResourcesTableProps = {
  Table: ConfiguredDataTable,
  loader?: JSX.Element,
  response: Response,
  resourceName: string,
};

const renderContent = ({
  Table,
  loader,
  resourceName,
  response,
}: {
  Table: ConfiguredDataTable,
  loader?: JSX.Element,
  resourceName: string,
  response: Response,
}): JSX.Element => {
  const {
    isLoading,
    hasData,
  } = response;

  if (isLoading) {
    return renderLoader({ loader, resourceName });
  }

  if (!hasData) {
    return (<Table data={{}} name={resourceName} />);
  }

  const data = response.data as DataTableData;

  return (<Table data={data} name={resourceName} />);
};

const renderLoader = ({
  loader,
  resourceName,
}: {
  loader?: JSX.Element,
  resourceName: string,
}): JSX.Element => {
  if (loader) { return(loader); }

  const message = `Loading ${titleCase(resourceName)}...`;

  return (<LoadingOverlay message={message} />);
};

export const ResourcesTable = ({
  Table,
  loader,
  response,
  resourceName,
}: ResourcesTableProps): JSX.Element => (
  <div className="min-h-[10rem] relative resources-table">
    { renderContent({ Table, loader, resourceName, response }) }
  </div>
);
