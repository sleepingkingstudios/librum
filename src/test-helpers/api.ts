import fetchMock from 'jest-fetch-mock';
import {
  Middleware,
  Reducer,
  configureStore,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query/react';

import {
  actions,
  reducer as session,
} from '@session';
import type { User } from '@session';
import type {
  ApiParam,
  ApiResponse,
} from '@api';

type Dispatch = (arg: unknown) => unknown;

type Endpoint = { initiate: (param: unknown) => unknown };

type GetDefaultMiddleware = () => Middleware[];

type Response = {
  data?: unknown;
  error?: {
    error: string;
  },
  status: 'fulfilled' | 'rejected';
}

type Store = ReturnType<typeof configureStore>;

interface ApiService {
  endpoints: {
    [name: string]: Endpoint;
  };
  middleware: Middleware;
  reducer: Reducer;
}

interface DispatchRequest {
  api: ApiService;
  endpoint: string;
  param?: ApiParam;
  store?: Store;
}

interface ShouldMatchTheRequest {
  headers?: { [name: string]: string };
  method?: string;
  url: string;
}

interface ShouldPerformTheMutation {
  api: ApiService;
  data: ApiResponse;
  endpoint: string;
  param?: ApiParam;
  request: ShouldMatchTheRequest;
}

interface ShouldPerformTheQuery {
  api: ApiService;
  data: ApiResponse;
  endpoint: string;
  param?: ApiParam;
  request: ShouldMatchTheRequest;
}

const authenticatedRequest = (
  { request, token }: { request: ShouldMatchTheRequest, token: string }
) => {
  const { headers, method, url } = request;
  const authenticatedHeaders = {
    ...headers,
    'Authorization': `Bearer ${token}`,
  };

  return {
    headers: authenticatedHeaders,
    method,
    url,
  };
}

const createApiStore = ({ api }: { api: ApiService }): Store => {
  const middleware = (
    getDefaultMiddleware: GetDefaultMiddleware,
  ) => getDefaultMiddleware().concat(api.middleware);
  const reducer = {
    api: api.reducer,
    session,
  };
  const store = configureStore({
    middleware,
    reducer,
  });

  setupListeners(store.dispatch);

  return store;
};

const dispatchRequest = async (
  {
    api,
    endpoint,
    param,
    store,
  }: DispatchRequest
) => {
  let configuredStore = store;

  if (!store) { configuredStore = createApiStore({ api }); }

  const dispatch: Dispatch = configuredStore.dispatch;

  const { endpoints } = api;
  const query = endpoints[endpoint];

  return await dispatch(query.initiate(param)) as Response;
};

const shouldMatchTheRequest = (
  { method = 'GET', headers = {}, url }: ShouldMatchTheRequest
) => {
  const request = fetchMock.mock.calls[0][0];

  if (typeof request === 'string') { throw(`Request??? ${request}`); }

  expect(request.method).toBe(method.toUpperCase());
  expect(request.url).toBe(url);

  expect(request.headers.get('Accept')).toBe('application/json');

  Object.entries(headers).forEach(([header, value]) => {
    expect(request.headers.get(header)).toBe(value);
  });
};

// eslint-disable-next-line jest/no-export
export const shouldPerformTheMutation = (
  {
    api,
    data,
    endpoint,
    param,
    request,
  }: ShouldPerformTheMutation
) => {
  describe('should perform the mutation', () => {
    // eslint-disable-next-line jest/expect-expect
    it('should update the data', async () => {
      await dispatchRequest({ api, endpoint, param });

      shouldMatchTheRequest(request);
    });

    describe('with a successful response', () => {
      it('should fulfill the request', async () => {
        fetchMock.once(JSON.stringify(data));

        const response = await dispatchRequest({ api, endpoint });

        expect(response.data).toEqual(data);
      });
    });

    describe('with a failing response', () => {
      it('should reject the request', async () => {
        fetchMock.mockRejectOnce(new Error('Something went wrong'));

        const response = await dispatchRequest({ api, endpoint });

        expect(response.error.error).toBe('Error: Something went wrong');
      });
    });

    describe('when the user is authenticated', () => {
      const { create } = actions;
      const user: User = {
        email: 'alan.bradley@example.com',
        id: '00000000-0000-0000-0000-000000000000',
        role: 'user',
        slug: 'alan-bradley',
        username: 'Alan Bradley',
      };
      const token = '12345';

      // eslint-disable-next-line jest/expect-expect
      it('should update the data', async () => {
        const store = createApiStore({ api });
        const dispatch: Dispatch = store.dispatch;
        const action = create({ token, user });

        dispatch(action);

        await dispatchRequest({
          api,
          endpoint,
          param,
          store,
        });

        shouldMatchTheRequest(authenticatedRequest({ request, token }));
      });
    });
  });
};

// eslint-disable-next-line jest/no-export
export const shouldPerformTheQuery = (
  {
    api,
    data,
    endpoint,
    param,
    request,
  }: ShouldPerformTheQuery
) => {
  describe('should perform the query', () => {
    // eslint-disable-next-line jest/expect-expect
    it('should query the data', async () => {
      await dispatchRequest({ api, endpoint, param });

      shouldMatchTheRequest(request);
    });

    describe('with a successful response', () => {
      it('should fulfill the request', async () => {
        fetchMock.once(JSON.stringify(data));

        const response = await dispatchRequest({ api, endpoint });
        const { status } = response;

        expect(response.data).toEqual(data);
        expect(status).toBe('fulfilled');
      });
    });

    describe('with a failing response', () => {
      it('should reject the request', async () => {
        fetchMock.mockRejectOnce(new Error('Something went wrong'));

        const response = await dispatchRequest({ api, endpoint });
        const { error: { error }, status } = response;

        expect(error).toBe('Error: Something went wrong');
        expect(status).toBe('rejected');
      });
    });

    describe('when the user is authenticated', () => {
      const { create } = actions;
      const user: User = {
        email: 'alan.bradley@example.com',
        id: '00000000-0000-0000-0000-000000000000',
        role: 'user',
        slug: 'alan-bradley',
        username: 'Alan Bradley',
      };
      const token = '12345';

      // eslint-disable-next-line jest/expect-expect
      it('should query the data', async () => {
        const store = createApiStore({ api });
        const dispatch: Dispatch = store.dispatch;
        const action = create({ token, user });

        dispatch(action);

        await dispatchRequest({
          api,
          endpoint,
          param,
          store,
        });

        shouldMatchTheRequest(authenticatedRequest({ request, token }));
      });
    });
  });
};
