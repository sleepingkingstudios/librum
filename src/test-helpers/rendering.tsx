import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
