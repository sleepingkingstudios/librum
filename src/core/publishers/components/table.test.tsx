import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { PublishersTable } from './table';
import { publishers } from '../test-helpers';

describe('<PublishersTable />', () => {
  const expectedHeader = [
    'Name',
    'Website',
    ' ',
  ];

  describe('when there is no matching data', () => {
    const data = {};
    const expectedMessage = 'There are no matching publishers.';

    it('should render the header cells', () => {
      const { getAllByRole } = render(<PublishersTable data={data} />);
      const cells = getAllByRole('columnheader');

      expect(cells).toHaveLength(3);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
    });

    it('should display the empty message', () => {
      const { getByRole } = render(<PublishersTable data={data} />);
      const cell = getByRole('cell');

      expect(cell.textContent).toBe(expectedMessage);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<PublishersTable data={data} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are many rows', () => {
    const data = { publishers };
    const expectedBody = [
      'Crystal Sphere Publications',
      'www.example.com/crystal-sphere-publications',
      ' ',
      'Flumph Free Press',
      'www.example.com/flumph-free-press',
      ' ',
      'Spelljammer Monthly',
      'www.example.com/spelljammer-monthly',
      ' ',
    ];

    it('should render the header cells', () => {
      const { getAllByRole } = render(
        <PublishersTable data={data} />,
        { router: true },
      );
      const cells = getAllByRole('columnheader');

      expect(cells).toHaveLength(3);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
    });

    it('should display the formatted data', () => {
      const { getAllByRole } = render(
        <PublishersTable data={data} />,
        { router: true },
      );
      const cells = getAllByRole('cell');

      expect(cells).toHaveLength(9);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedBody);
    });

    it('should render the name links', () => {
      const { getByRole } = render(
        <PublishersTable data={data} />,
        { router: true },
      );

      const link = getByRole('link', { name: 'Crystal Sphere Publications' });
      expect(link).toBeVisible();
      expect(link).toHaveAttribute(
        'href',
        '/publishers/crystal-sphere-publications',
      );
    });

    it('should render the website links', () => {
      const { getByRole } = render(
        <PublishersTable data={data} />,
        { router: true },
      );

      const link = getByRole(
        'link',
        { name: 'www.example.com/crystal-sphere-publications' },
      );
      expect(link).toBeVisible();
      expect(link).toHaveAttribute(
        'href',
        'https://www.example.com/crystal-sphere-publications',
      );
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PublishersTable data={data} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
