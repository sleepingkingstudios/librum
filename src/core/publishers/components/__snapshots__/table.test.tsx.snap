// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublishersTable /> when there are many rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-collapse data-table-core-publishers"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-name"
          >
            Name
          </th>
          <th
            class="data-table-header-website"
          >
            Website
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-cell-name"
            data-label="Name"
          >
            <a
              href="/publishers/crystal-sphere-publications"
            >
              Crystal Sphere Publications
            </a>
          </td>
          <td
            class="data-table-cell-website"
            data-label="Website"
          >
            <a
              href="https://www.example.com/crystal-sphere-publications"
            >
              <span>
                www.example.com/crystal-sphere-publications
              </span>
            </a>
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
             
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-name"
            data-label="Name"
          >
            <a
              href="/publishers/flumph-free-press"
            >
              Flumph Free Press
            </a>
          </td>
          <td
            class="data-table-cell-website"
            data-label="Website"
          >
            <a
              href="https://www.example.com/flumph-free-press"
            >
              <span>
                www.example.com/flumph-free-press
              </span>
            </a>
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
             
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-name"
            data-label="Name"
          >
            <a
              href="/publishers/spelljammer-monthly"
            >
              Spelljammer Monthly
            </a>
          </td>
          <td
            class="data-table-cell-website"
            data-label="Website"
          >
            <a
              href="https://www.example.com/spelljammer-monthly"
            >
              <span>
                www.example.com/spelljammer-monthly
              </span>
            </a>
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
             
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<PublishersTable /> when there is no matching data should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-collapse data-table-core-publishers"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-name"
          >
            Name
          </th>
          <th
            class="data-table-header-website"
          >
            Website
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="3"
          >
            There are no matching publishers.
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;
